spec.bs

```diff
@@ -243,8 +250,200 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
   1. Return [=this=]'s [=URLPattern/hash component=]'s [=component/pattern string=].
 </div>
 
+<div algorithm>
+  The <dfn method for="URLPattern">test(|input|, |baseURL|)</dfn> method steps are:
+
+  1. Let |result| be the result of [=match=] given [=this=], |input|, and |baseURL|.
+  1. If |result| is null, return false.
+  1. Return true.
+</div>
+
+<div algorithm>
+  The <dfn method for="URLPattern">exec(|input|, |baseURL|)</dfn> method steps are:
+
+  1. Return the result of [=match=] given [=this=], |input|, and |baseURL|.
+</div>
+
+<div algorithm>
+  To perform a <dfn>match</dfn> given a {{URLPattern}} |urlpattern|, a {{URLPatternInput}} |input|, and an optional string |baseURLString|:
+
+  1. Let |protocol| be the empty string.
+  1. Let |username| be the empty string.
+  1. Let |password| be the empty string.
+  1. Let |hostname| be the empty string.
+  1. Let |port| be the empty string.
+  1. Let |pathname| be the empty string.
+  1. Let |search| be the empty string.
+  1. Let |hash| be the empty string.
+  1. Let |inputs| be an empty array.
+  1. Push |input| on to |inputs|.
+  1. If |input| is a {{URLPatternInit}} then:
+    1. If |baseURLString| is not null, throw a TypeError.
+    1. Let applyResult be the result of [=process a URLPatternInit=] given |input|, "url", |protocol|, |username|, |password|, |hostname|, |port|, |pathname|, |search|, and |hash|. If this throws an exception, catch it, and return null.
+    1. Set |protocol| to |applyResult|'s {{URLPatternInit/protocol}}.
+    1. Set |username| to |applyResult|'s {{URLPatternInit/username}}.
+    1. Set |password| to |applyResult|'s {{URLPatternInit/password}}.
+    1. Set |hostname| to |applyResult|'s {{URLPatternInit/hostname}}.
+    1. Set |port| to |applyResult|'s {{URLPatternInit/port}}.
+    1. Set |pathname| to |applyResult|'s {{URLPatternInit/pathname}}.
+    1. Set |search| to |applyResult|'s {{URLPatternInit/search}}.
+    1. Set |hash| to |applyResult|'s {{URLPatternInit/hash}}.
+  1. Else:
+    1. Let |baseURL| be null.
+    1. If |baseURLString| is not null then:
+      1. Set |baseURL| to the result of [=URL parser|parsing=] |baseURLString|.
+      1. If |baseURL| is failure, return null.
+      1. Push |baseURL| on to |inputs|.
+    1. Let |url| be the result of [=URL parser|parsing=] |input| given |baseURL|.
+      1. If |url| is failure, return null.
+    1. Set |protocol| to |url|'s {{URL/protocol}}.
```

Comment:
It's not great practice to refer directly to object properties like `protocol`, since they are modifiable by JavaScript (e.g. JS messing with `URL.prototype`).

Additionally, |url| is a [URL record](https://url.spec.whatwg.org/#concept-url), not a `URL` instance.

This is tricky. I see a few options:

1. Duplicate all the getter steps logic here for the components you're interested in.
2. Work on a refactoring of the URL spec to abstract these all out into algorithms you can call, e.g. "the API hostname string" for a URL record.
3. Try to sidestep the issue by creating Web IDL objects and referencing them.

(3) is somewhat unprecedented but perhaps easiest. It would look something like:

```
1. Let |urlObject| be a [=new=] {{URL}}.
2. Set |urlObject|'s [=URL/URL=] to |url|.
3. Set |protocol| to the result of applying the getter steps for {{URL/protocol}} to |urlObject|.
```

Having written this out I do not like this very much. I would suggest (2). What that would look like for now would be adding a section at the bottom of your spec which says something like

```html
This spec depends on factoring out various getter steps for the {{URL}} object into new algorithms that operate on URL records and return [=scalar value strings=]. These are the <dfn for="url">API protocol string</a>, the <dfn for="url">API username string</a>, ...
```

i.e. you don't have to do the work of actually factoring things out at this stage; just rely on it being rather obvious.

Reply:
For a lot of these I actually don't want the getter behavior anyway.  For example, I just want the scheme and not the trailing `:` added.  So I think I can reference the record values directly.  The only one I probably need to get exported is the pathname getter since it does extra work to construct the returned string with slashes, etc.